File: leetcode/710. 黑名单中的随机数.cs
# 710. 黑名单中的随机数
# https://leetcode.cn/problems/random-pick-with-blacklist/
public class Solution
{
    readonly int whiteRange;
    readonly Random random = new();
    readonly Dictionary<int, int> map;

    public Solution(int n, int[] blacklist)
    {
        whiteRange = n - blacklist.Length;
        var blackset = blacklist.ToHashSet();
        var blackInRange = blacklist
            .Where(x => x < whiteRange);
        //var blackInRange = Range(0, whiteRange)
        //    .Where(blackset.Contains);
        var whiteNotInRange = Range(whiteRange, n)
            .Where(x => !blackset.Contains(x));
        map = blackInRange
            .Zip(whiteNotInRange)
            .ToDictionary(p => p.First, p => p.Second);
    }

    public int Pick()
    {
        int x = random.Next(0, whiteRange);
        return map.GetValueOrDefault(x, x);
    }

    static IEnumerable<int> Range(int start, int stop)
    {
        for (int i = start; i < stop; i++)
            yield return i;
    }
}

class LeetCode
{
    public static void Main()
    {
        var sol = new Solution(2, Array.Empty<int>());
        for (int c = 1; c < 10; c++)
            Console.WriteLine(sol.Pick());
    }
}

File: leetcode/710. 黑名单中的随机数.py
# 710. 黑名单中的随机数
# https://leetcode.cn/problems/random-pick-with-blacklist/
from typing import List, Dict, Set, Tuple, Iterable, Optional
import math
import random


class Solution:
    def __init__(self, n: int, blacklist: List[int]):
        self.white_range = n - len(blacklist)
        blackset = set(blacklist)
        # black_in_range = filter(lambda x: x in blackset, range(0, self.white_range))
        black_in_range = filter(lambda x: x < self.white_range, blacklist)
        white_not_in_range = filter(lambda x: x not in blackset, range(self.white_range, n))
        self.map = dict(zip(black_in_range, white_not_in_range))

    def pick(self) -> int:
        x = random.randint(0, self.white_range - 1)
        return self.map.get(x, x)


if __name__ == '__main__':
    pass
